tsl/src/compression/recompress.c

```diff
@@ -0,0 +1,725 @@
+	Oid *sort_operators = palloc(sizeof(*sort_operators) * n_keys);
+	Oid *sort_collations = palloc(sizeof(*sort_collations) * n_keys);
+	bool *nulls_first = palloc(sizeof(*nulls_first) * n_keys);
+
+	CompressedSegmentInfo *current_segment = palloc0(sizeof(CompressedSegmentInfo) * n_keys);
+
+	for (int n = 0; n < n_keys; n++)
+	{
+		const char *attname;
+		int position;
+		if (n < num_segmentby)
+		{
+			position = n + 1;
+			attname = ts_array_get_element_text(settings->fd.segmentby, position);
+		}
+		else
+		{
+			position = n - num_segmentby + 1;
+			attname = ts_array_get_element_text(settings->fd.orderby, position);
+		}
+
+		AttrNumber col_attno = get_attnum(uncompressed_chunk_rel->rd_id, attname);
+		current_segment[n].decompressed_chunk_offset = AttrNumberGetAttrOffset(col_attno);
+		current_segment[n].segment_info = segment_info_new(
+			TupleDescAttr(uncompressed_rel_tupdesc, current_segment[n].decompressed_chunk_offset));
+
+		compress_chunk_populate_sort_info_for_column(settings,
+													 RelationGetRelid(uncompressed_chunk_rel),
+													 attname,
+													 &sort_keys[n],
+													 &sort_operators[n],
+													 &sort_collations[n],
+													 &nulls_first[n]);
+	}
+
+	/******************** row decompressor **************/
+
+	RowDecompressor decompressor = build_decompressor(compressed_chunk_rel, uncompressed_chunk_rel);
+	/********** row compressor *******************/
+	RowCompressor row_compressor;
+	row_compressor_init(settings,
+						&row_compressor,
+						uncompressed_chunk_rel,
+						compressed_chunk_rel,
+						compressed_rel_tupdesc->natts,
+						true /*need_bistate*/,
+						0 /*insert options*/);
+
+	Relation index_rel = index_open(row_compressor.index_oid, ExclusiveLock);
+	ereport(DEBUG1,
+			(errmsg("locks acquired for recompression: \"%s.%s\"",
+					NameStr(uncompressed_chunk->fd.schema_name),
+					NameStr(uncompressed_chunk->fd.table_name))));
+
+	/* Setting up scankeys */
+	ScanKeyData *index_scankeys = palloc(sizeof(ScanKeyData) * num_segmentby);
+	ScanKeyData *orderby_scankeys = palloc(sizeof(ScanKeyData) * num_orderby * 2);
+	create_segmentby_scankeys(settings, index_rel, compressed_chunk_rel, index_scankeys);
+	create_orderby_scankeys(settings, index_rel, compressed_chunk_rel, orderby_scankeys);
+
+	Tuplesortstate *segment_tuplesortstate = tuplesort_begin_heap(uncompressed_rel_tupdesc,
+																  n_keys,
+																  sort_keys,
+																  sort_operators,
+																  sort_collations,
+																  nulls_first,
+																  maintenance_work_mem,
+																  NULL,
+																  false);
+
+	Tuplesortstate *batch_tuplesortstate = tuplesort_begin_heap(uncompressed_rel_tupdesc,
+																n_keys,
+																sort_keys,
+																sort_operators,
+																sort_collations,
+																nulls_first,
+																maintenance_work_mem,
+																NULL,
+																false);
```

Comment:
This is always inside one segment group, should it sort only by orderby columns?

The name is not very precise, maybe something like `recompressing_tuplesort`? In the sense that these are the tuples of the batches that we're currently recompressing.

Reply:
Renamed to `recompress_tuplesortstate` and added a comment for clearing up the usage pattern.